### ISSUE TITLE
DefaultFSTInt2ObjectMap clear issue leads to JVM crash

### ISSUE THREAD

**Issue by titonio** [RuedigerMoeller/fast-serialization#278]
Using 2.57 version taken from maven central (it seems there's no tag here in git) we were encountering issues during deserialization which either caused a Properties object to report incorrect values or ocasionally JVM crash.

After further tests and inspection we narrowed the issue to an incorrect clean of the DefaultFSTInt2ObjectMap, leaving some values which should be removed.

We created a simple unit test to reproduce the issue

```
    @Test
    public void testIncompleteClear() {

        DefaultFSTInt2ObjectMap<String> map = new DefaultFSTInt2ObjectMap<>(6);

        map.put(1, "first");
        map.put(14, "second");

        assertEquals("first", map.get(1));

        map.clear();

        map.put(19, "third");

        assertNull("Should be null", map.get(14));
    }
```

As you can see the next was not correctly cleared in this case.

This may be related with #274  and #235

**Comment by RuedigerMoeller** [comment]
thanks for the report .. wonder how this relates to a crash though. Might try to run it with unsafe disabled in order to get a proper exception.

**Comment by benjamingouit** [comment]
I have the same problem using 2.56 version.
It appears that some values from previous deserialization are kept in org.nustaq.serialization.FSTObjectRegistry.idToObject (which is a DefaultFSTInt2ObjectMap).
Moreover, the code in org.nustaq.serialization.FSTObjectRegistry::getReadRegisteredObject returns object from this idToObject map before the object in reuseMap :
```java
    public Object getReadRegisteredObject(int handle) {
        if (disabled) {
            return null;
        }
        int pos = handle / OBJ_DIVISOR;
        if ( pos < reuseMap.length ) {
            if ( reuseMap[pos] == null ) {
                return null;
            } else {
                Object candidate = idToObject.get(handle);
                if ( candidate == null )
                    return reuseMap[pos];
                return candidate;
            }
        }
        return idToObject.get(handle);
    }
```

Unfortunately the value at the same position in reuseMap is the good one.
Because of that, a key in the deserialized map had changed which is causing a NullPointerException later.

We are deserializing by calling this :
```java
public static final ThreadLocal<FSTConfiguration> FST_CONFIGURATION = ThreadLocal
            .withInitial(FSTConfiguration::createDefaultConfiguration);

        byte[] message;
        FST_CONFIGURATION.get().asObject(message);
```

I also tried with :
```java
        byte[] message;
        FSTConfiguration.getDefaultConfiguration().asObject(message);
```

So I have some questions about this :
- Is there a way to really clear this map ? Or a way to get a clean newly configuration ?

- Why there is a map (idToObject) and an array (reuseMap) and why the map value is chosen rather than the array value ?

**Comment by Arglanir** [comment]
It think it is simply related to the `clear()` code that clears the `next` map only if `mKeys.length > 6 * size() && size > 0` is false. It should instead always be done.
```java
public void clear() {
        int size = size();
        if (size == 0)
            return;
        if (mKeys.length > 6 * size() && size > 0) {
            // avoid cleaning huge mem areas after having written a large object
            if (size < 2) {
                size = 2;
            }
            size = FSTObject2IntMap.adjustSize(size * 2);
            mKeys = new int[size];
            mValues = new Object[size];
            mNumberOfElements = 0;
        } else {
            FSTUtil.clear(mKeys);
            FSTUtil.clear(mValues);
            mNumberOfElements = 0;
            if (next != null) {
                next.clear();
            }
        }
    }
```
Should be:
```java
public void clear() {
        int size = size();
        if (size == 0)
            return;
        if (mKeys.length > 6 * size() && size > 0) {
            // avoid cleaning huge mem areas after having written a large object
            if (size < 2) {
                size = 2;
            }
            size = FSTObject2IntMap.adjustSize(size * 2);
            mKeys = new int[size];
            mValues = new Object[size];
            mNumberOfElements = 0;
        } else {
            FSTUtil.clear(mKeys);
            FSTUtil.clear(mValues);
            mNumberOfElements = 0;
        }
        if (next != null) {
            next.clear();
        }
    }
```

**Comment by RuedigerMoeller** [comment]
fixed in 3.0.1
